src/utils/body-limit.ts

```diff
@@ -0,0 +1,86 @@
+import type { H3Event } from "../event.ts";
+import type { H3Plugin } from "../types/h3.ts";
+import { HTTPError } from "../error.ts";
+import { defineMiddleware } from "../middleware.ts";
+
+export interface BodySizeLimitOptions {
+  /**
+   * Maximum allowed body size in bytes
+   */
+  maxSize: number;
+  /**
+   * Routes to apply the limit to (optional)
+   * If not specified, applies to all routes
+   */
+  routes?: Array<string | RegExp>;
+  /**
+   * Routes to exclude from the limit (optional)
+   */
+  exclude?: Array<string | RegExp>;
+}
+
+/**
+ * Define a plugin that limits request body size
+ *
+ * @example
+ * ```js
+ * import { defineBodySizeLimitPlugin } from "h3";
+ *
+ * const bodySizeLimit = defineBodySizeLimitPlugin({
+ *   maxSize: 1024 * 1024, // 1MB
+ *   routes: ["/api/upload", /^\/api\/files/],
+ *   exclude: ["/api/large-upload"]
+ * });
+ *
+ * app.register(bodySizeLimit);
+ * ```
+ */
+export function defineBodySizeLimitPlugin(
+  options: BodySizeLimitOptions,
+): H3Plugin {
+  return (h3) => {
+    h3.use(
+      defineMiddleware(async (event: H3Event) => {
+        const url = event.req.url;
+        const path = url ? new URL(url, "http://localhost").pathname : "/";
```

Comment:
Creating a new `URL` object for every request is computationally expensive - it involves parsing the entire URL structure, validating components, and creating multiple internal objects.
This operation can be 5–10x slower than simple string operations when you're only trying to extract the pathname.

You can optimize it by using the utility from `srvx`:
[https://github.com/h3js/srvx/blob/main/src/\_url.ts](https://github.com/h3js/srvx/blob/main/src/_url.ts)
or just go with something like:

```ts
event.req.url?.split('?')[0] || '/'
```

Reply:
https://github.com/h3js/h3/pull/1150/files#r2189726225, middleware itself should not add extra logic for route matching